### PR TITLE
Hyperparameters

### PR DESCRIPTION
**Pull request by alfredofernandez2021**
Hyperparameter inclusions and memory management

### CODE REVIEW

**Comment by alfredofernandez2021** [comment]
Welcome to the Main Menu!
1) Create Neural Network
2) Load Neural Network
3) Introduction and Info
4) Exit Network Manager
Selection: 1

Creation:
What is the length of inputs that this neural network will accept? 784

What is the number of outputs that this neural network will produce? 10

How many layers will this neural network contain? 2

Which cost function should be used to calculate error? 0

What is the learning rate of this network? 0.00000001

What is the learning decay of this network? 0

What is the batch size that this network will train on? 1

What are the total epochs that this network will train on? 1

What is the momentum retention of each training batch? 0

What is the percent neuron dropout of the network in each batch/epoch? 0

What is the minimun network error needed to prevent learning on a sample? 100

What is the average maximum percent error that will conclude training? 0


Define neural layer 2:
        Activation type: 0


Manage:
1) Select DataSets
2) Run Training
3) Run Testing
4) Save Solution
5) Help
6) Back
Selection: 4

Save:
Enter name of file to save network as: new1.xml

Manage:
1) Select DataSets
2) Run Training
3) Run Testing
4) Save Solution
5) Help
6) Back
Selection: 6

Welcome to the Main Menu!
1) Create Neural Network
2) Load Neural Network
3) Introduction and Info
4) Exit Network Manager
Selection: 2

Loading:
Enter XML file name to load from: new1.xml

Manage:
1) Select DataSets
2) Run Training
3) Run Testing
4) Save Solution
5) Help
6) Back
Selection: 1

Dataset:
Training set image file path: a
Training set label file path: a
Testing set image file path: a
Testing set label file path: a

Manage:
1) Select DataSets
2) Run Training
3) Run Testing
4) Save Solution
5) Help
6) Back
Selection: 2

Training:
Batch size: 1
Learning rate: 1
Current score: 0.12
answer: 3       correct: 5

Current score: 0.12
answer: 2       correct: 1

Current score: 0.11
answer: 2       correct: 8

Current score: 0.105
answer: 2       correct: 2

...

Current score: 0.101809
answer: 2       correct: 4

Current score: 0.101806
answer: 2       correct: 5

Current score: 0.10177
answer: 2       correct: 4

Final score: 0.101717
Type 0 to exit:
0

Manage:
1) Select DataSets
2) Run Training
3) Run Testing
4) Save Solution
5) Help
6) Back
Selection: 4

Save:
Enter name of file to save network as: new2.xml

Manage:
1) Select DataSets
2) Run Training
3) Run Testing
4) Save Solution
5) Help
6) Back
Selection: 6

Welcome to the Main Menu!
1) Create Neural Network
2) Load Neural Network
3) Introduction and Info
4) Exit Network Manager
Selection: 4

**Comment by alfredofernandez2021** [comment]
image

Memory management - NeuralNetworks are now being deallocated at start of mainSelection() FSM state (after 'back')

**Comment by alfredofernandez2021** [comment]
Purpose of this PR is to implement and use all hyperparameters to mark the start of alpha version

**Comment by alfredofernandez2021** [comment]
784 10 3 1 0.0001 0 1 1 0 0 1 0 2 20 3. train, test, save, back, load, test, train -> cases of all overflown output at -inf

**Comment by alfredofernandez2021** [comment]
784 10 3 1 0.0001 0 1 1 0 0 1 0 2 20 3 incorrect output layer length. Another saving/loading issue.

**Comment by alfredofernandez2021** [comment]
"Saving layerType" commit fixed 784 10 3 1 0.0001 0 1 1 0 0 1 0 2 20 3 load->train exploding output problem. Try with all linears next

**Comment by alfredofernandez2021** [comment]
PR pre-merge goals:
   Implement: softmax layer, exponential layer, binary layer
   Fix: Learning issue for linear neurons
   Verify: Linear networks still work after being saved and loaded

**Comment by alfredofernandez2021** [comment]
Todo: 
Implement Binary, Exponential, Softmax Neurons into menu flow
Test Binary, Exponential, Softmax Neuron layers
Fix learning in linear layer (likely underflow and overflow taking place in specialized neurons)

**Comment by alfredofernandez2021** [comment]
784 10 2 1 0.0001 0 1 1 0 0 1 0 1

Current score: 0.15828
maxValue: -0.862122     answer: 7       correct: 4

Final score: 0.1584

Type 0 to exit:

**Comment by alfredofernandez2021** [comment]
784 10 2 1 0.0001 0 1 1 0 0 1 0 1

Current score: 0.425826
maxValue: -3.48491      answer: 1       correct: 4

Final score: 0.426017

Type 0 to exit:

**Comment by alfredofernandez2021** [comment]
784 10 2 1 0.0001 0 1 1 0 0 1 0 1

Current score: 0.610451
maxValue: -3.40451      answer: 1       correct: 4

Final score: 0.6106

Type 0 to exit:

**Comment by alfredofernandez2021** [comment]
784 10 2 1 0.0001 0 1 1 0 0 1 0 1

Current score: 0.659649
maxValue: -3.10385      answer: 1       correct: 4

Final score: 0.659733

Type 0 to exit:

**Comment by alfredofernandez2021** [comment]
784 10 2 1 0.0001 0 1 1 0 0 1 0 1

Current score: 0.747963
maxValue: -2.77487      answer: 4       correct: 4

Final score: 0.747967

Type 0 to exit:
